src/sage/groups/additive_abelian/additive_abelian_wrapper.py

```diff
@@ -566,3 +616,188 @@ def _rec(j, k, c):
+#    assert not sum(r*a for r,a in zip(rel, alphas+[beta]))
+#    assert all(a.order() == p**v for a,v in zip(alphas,vals))
+
+    # step 1
+    for i in range(k-1):
+        if not rel[i]:
+            continue
+        q = rel[i].p_primary_part(p)
+        alphas[i] *= rel[i]//q
+        rel[i] = q
+    min_r = min(filter(bool, rel))
+    val_rlast = rel[-1].valuation(p)
+    assert rel[-1] == p**val_rlast
+#    assert not sum(r*a for r,a in zip(rel, alphas+[beta]))
+
+    # step 2
+    if rel[-1] == min_r:
+        for i in range(k-1):
+            beta += alphas[i] * (rel[i]//rel[-1])
+        alphas.append(beta)
+        vals.append(val_rlast)
+#        assert alphas[-1].order() == p**vals[-1]
+        return
+
+    # step 3
+    j = next(j for j,r in enumerate(rel) if r == min_r)
+    alphas[j] = sum(a * (r//rel[j]) for a,r in zip(alphas+[beta], rel))
+
+    # step 4
+    if not alphas[j]:
+        del alphas[j], vals[j]
+        if not alphas:
+            alphas.append(beta)
+            vals.append(val_rlast)
+#            assert alphas[-1].order() == p**vals[-1]
+            return
+
+    # step 5
+    beta_q = beta
+    for v in range(1, h):
+        beta_q *= p
+        try:
+            e = _discrete_log_pgroup(p, vals, alphas, -beta_q)
+        except TypeError:
+            continue
+        else:
+            # step 6
+            _expand_basis_pgroup(p, alphas, vals, beta, h, list(e) + [p**v])
+#            assert alphas[-1].order() == p**vals[-1]
+            return
+    else:
+        alphas.append(beta)
+        vals.append(h)
+#        assert alphas[-1].order() == p**vals[-1]
+
+def basis_from_generators(gens, ords=None):
+    r"""
+    Given a generating set of some finite abelian group
+    (additively written), compute and return a basis of
+    the group.
+
+    .. NOTE::
+
+        A *basis* of a finite abelian group is a generating
+        set `\{g_1,...,g_n\}` such that each element of the
+        group can be written as a unique linear combination
+        `\alpha_1 g_1 + ... + \alpha_n g_n` with each
+        `\alpha_i \in \{0,...,\mathrm{ord}(g_i)-1\}`.
+
+    ALGORITHM: [Suth2007]_, Algorithm 9.1 & Remark 9.1
+
+    EXAMPLES::
+
+        sage: from sage.groups.additive_abelian.additive_abelian_wrapper import basis_from_generators
+        sage: E = EllipticCurve(GF(31337^6,'a'), j=37)
+        sage: E.order()
+        946988065073788930380545280
+        sage: (R,S), (ordR,ordS) = basis_from_generators(E.gens())
+        sage: ordR, ordS
+        (313157428926517503432720, 3024)
+        sage: R.order() == ordR
+        True
+        sage: S.order() == ordS
+        True
+        sage: ordR * ordS == E.order()
+        True
+        sage: R.weil_pairing(S, ordR).multiplicative_order() == ordS
+        True
+        sage: E.abelian_group().invariants()
+        (3024, 313157428926517503432720)
+    """
+    if ords is None:
+        ords = [g.order() for g in gens]
+    if not gens:
+        return [], []
```

Comment:
```suggestion
    if not gens:
        return [], []
    if ords is None:
        ords = [g.order() for g in gens]
```
Not that it makes that much of a difference, but it is one less check for empty `gens`.